format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala

```diff
@@ -282,6 +268,112 @@ object SyntaxTokens {
     def tokensWith: KwWith = tree.findBetween[KwWith](_.lhs, _.rhs).get
   }
 
+  // == Defn ==
+  implicit class XtensionDefnClassSyntax(val tree: Defn.Class) extends AnyVal {
+    def tokensClass: KwClass = tree.find[KwClass].get
+
+    // tparams
+    def tokensLeftBracket: Option[LeftBracket] =
+      tree.tparams.headOption
+        .map(tparam => tree.findBetween[LeftBracket](_.name, _ => tparam).get)
+    def tokensRightBracket: Option[RightBracket] =
+      tree.tparams.lastOption.map(
+        tparam =>
+          tree.ctor.paramss match {
+            case Nil => tree.findAfter[RightBracket](_ => tparam).get
+            case xs :: _ =>
+              xs match {
+                case Nil => tree.findAfter[RightBracket](_ => tparam).get
+                case x :: _ =>
+                  tree.findBetween[RightBracket](_ => tparam, _ => x).get
+              }
+        }
+      )
+
+    def tokensCommaTparams: List[Comma] =
+      commaSeparated(tree)(_.tparams)
+
+    def tokensCommaCtor: List[List[Comma]] =
+      tree.ctor.paramss.map(commaSeparated0(tree))
+
+    def tokensImplicit: Option[KwImplicit] = {
+      val paramss = tree.ctor.paramss
+      if (paramss.isEmpty) None
+      else {
+        paramss.last match {
+          case p :: _ if p.mods.exists(_.is[Mod.Implicit]) => {
+            val kw =
+              tree.tokens
+                .leadings(p.tokens.head)
+                .find(_.is[KwImplicit])
+                .get
+                .asInstanceOf[KwImplicit]
+            Some(kw)
+          }
+          case _ => None
+        }
+      }
+    }
+
+    def tokensParenthesis: List[(LeftParen, RightParen)] = {
+      val paramss = tree.ctor.paramss
+      if (paramss.isEmpty) Nil
+      else {
+        val buf = List.newBuilder[(Token.LeftParen, Token.RightParen)]
+        val tokens = tree.tokens
+        val matching = MatchingParens(tokens)
```

Comment:
This is an expensive operation which should ideally only be done once per Source. Performance is not a big concern at the moment so this is fine but eventually I imagine this values becomes a method argument